### PR TITLE
Remove usage of Module.getCompilation

### LINKED ISSUE #28603: Implement DependencyGraph.getDirectDependents method
**Description:**
DependencyGraph.getDirectDependents method is supposed to return the list of modules that has the current module as a dependency. This should be used to clear caches of dependant modules when an edit operation is done. This could introduce a few more APIs along with it to retrieve transitive dependants as well.

The current way of clearing caches added in #28585 should be refactored once this method is implemented.

### PR DESCRIPTION
**Pull request by azinneera**
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by NipunaRanasinghe — COMMENTED** [review]
Debug server related changes look good to me.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/TestUtil.java [hunk 1]
```diff
@@ -540,10 +539,7 @@ public static String getResponseString(CompletableFuture completableFuture) {
         if (project.isEmpty()) {
             return diagnostics;
         }
-        for (Module module : project.get().currentPackage().modules()) {
-            diagnostics.addAll(module.getCompilation().diagnostics().diagnostics());
-        }
-
+        diagnostics.addAll(project.get().currentPackage().getCompilation().diagnosticResult().diagnostics());
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/TestUtil.java:542]: Shall we get this to a variable in order to improve the readability?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/references/ReferencesUtil.java [hunk 2]
```diff
@@ -81,9 +81,10 @@ private ReferencesUtil() {
         }
         
         Symbol symbol = symbolAtCursor.get();
-        project.get().currentPackage().modules().forEach(module -> {
-            List<Location> references = module.getCompilation().getSemanticModel().references(symbol);
-            moduleLocationMap.put(module, references);
+        project.get().currentPackage().moduleIds().forEach(moduleId -> {
+            List<Location> references = project.get().currentPackage()
+                    .getCompilation().getSemanticModel(moduleId).references(symbol);
+            moduleLocationMap.put(project.get().currentPackage().module(moduleId), references);
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/references/ReferencesUtil.java:87]: Get the module to a variable, in order to improve the readability here

File: misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/util/TestUtil.java [hunk 3]
```diff
@@ -210,10 +209,7 @@ public static String getResponseString(CompletableFuture completableFuture) {
         if (project.isEmpty()) {
             return diagnostics;
         }
-        for (Module module : project.get().currentPackage().modules()) {
-            diagnostics.addAll(module.getCompilation().diagnostics().diagnostics());
-        }
-
+        diagnostics.addAll(project.get().currentPackage().getCompilation().diagnosticResult().diagnostics());
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/util/TestUtil.java:212]: Shall we get this to a variable in order to improve the readability?

**Review by nadeeshaan — COMMENTED** [review]
LS related Changes LGTM

**Review by sameerajayasoma — CHANGES_REQUESTED** [review]
Let's make ModuleCompilation package private for now with this PR.

**Comment by azinneera** [comment]
> Let's make ModuleCompilation package private for now with this PR.

We couldn't completely remove it to make it package-private. It's still being used in TestProcessor and doc generation. These usages cannot be replaced as it is getting the semantic model after code gen.

File: misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/util/TestUtil.java (hunk 3 above)
Comment (azinneera) [reply misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/util/TestUtil.java]: Resolved in 1907cf8

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/references/ReferencesUtil.java (hunk 2 above)
Comment (azinneera) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/references/ReferencesUtil.java]: Resolved in 1907cf8

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/TestUtil.java (hunk 1 above)
Comment (azinneera) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/TestUtil.java]: Resolved in 1907cf8

**Review by sameerajayasoma — APPROVED** [review]

**Pull request merged**